### PR TITLE
SI-7982 make askLoadedType unload arguments out of the PC by default,

### PR DESCRIPTION
**Pull request by dotta**
Re-opened https://github.com/scala/scala/pull/3158 to target 2.10.x

review by @adriaanm

As point out by @huitseeker in https://github.com/scala/scala/pull/3158#issuecomment-28900230, scala-refactoring failure is expected until https://github.com/scala-ide/scala-refactoring/pull/37 is merged. I'm not sure what's the best course of action here, but merging this PR will require some coordination with @misto, because the scala-refactoring PR should be merged at the same time.

### CODE REVIEW

**Comment by misto** [comment]
Feel free to just merge it in scala-refactoring yourself once it's ready!

**Comment by dotta** [comment]
@misto Ok!

@adriaanm How do you want to coordinate. I can push the button on scala-refactoring side whenever you are ready.

**Comment by retronym** [comment]
We'll need the other tests to pass here, first :)

https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/3897/console

```
BUILD FAILED
/localhome/jenkins/c/workspace/pr-scala-test/scala/build.xml:1583: Test suite finished with 1 case failing:
/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/presentation/ide-t1001388 [FAILED]
```

**Comment by dotta** [comment]
> We'll need the other tests to pass here, first :)

Ouch. I'll look at it...

**Comment by dotta** [comment]
I'm going to have a dry-run at building the Scala IDE using this PR + https://github.com/scala-ide/scala-refactoring/pull/37, and report back in about 1h.

In the meanwhile, please, have a look at the changes and let me know if you want me to change anything. If it's good to ship, I'll squash the two commits and we can sync for merging this PR and scala-refactoring one at the same time. \cc  @retronym @adriaanm

**Comment by dotta** [comment]
I'm experiencing more trouble than I expected while rebuilding a Scala IDE on top of this PR. I'm going to close this PR so I won't be spamming you all while I'm working on fixing things.

**Comment by dotta** [comment]
Alright, this is now ready to be merged.

The following two PR have to be merged together, or the Scala IDE validator script will start failing:
- https://github.com/scala-ide/scala-ide/pull/594
- https://github.com/scala-ide/scala-refactoring/pull/37

I can push the merge button on the above two projects whenever you are ready.

**Comment by adriaanm** [comment]
@dotta, great! Since you control the merge buttons of those other projects, I suggest you hit them when you're ready and then do a rebuild here -- I'll merge later (my) today when the lights go green.

**Comment by adriaanm** [comment]
Is there any way you can structure the changes so that they work whether or not all PRs are merged at the same time? By having a fallback in the IDE patch, for example. I didn't look at the code, so I'm just asking.

**Comment by adriaanm** [comment]
Looking at the code a bit, I would (naively -- I haven't fully understood everything yet) suggest trying a different approach: make the `keepLoaded = true` behavior the default (this way scala-refactoring need not change) and offer an overload that doesn't keep stuff loaded (when/why should the argument be false?). In general, boolean arguments like this are a bit of a code smell.

**Comment by dotta** [comment]
> @dotta, great! Since you control the merge buttons of those other projects, I suggest you hit them when you're ready and then do a rebuild here -- I'll merge later (my) today when the lights go green.

@adriaanm It's a bit hard to coordinate with you due to the timezone, can someone from Europe be in charge of merging this PR?

> Is there any way you can structure the changes so that they work whether or not all PRs are merged at the same time? By having a fallback in the IDE patch, for example. I didn't look at the code, so I'm just asking.

I thought about that, but I don't see another way.

> Looking at the code a bit, I would (naively -- I haven't fully understood everything yet) suggest trying a different approach: make the keepLoaded = true behavior the default (this way scala-refactoring need not change) and offer an overload that doesn't keep stuff loaded (when/why should the argument be false?).  In general, boolean arguments like this are a bit of a code smell.

That would defeat the purpose of the PR, as the default should really be not to keep units loaded. The reason is that typechecking occurs every time the reconciler kicks-in (~500millis after you stop typing), and the more units are loaded the slower is the typechecking performed by the presentation compiler. Hence, we want callers of `askLoadedType` to think about keeping a unit loaded, which is why the defult is to not do so. [Here](https://www.assembla.com/spaces/scala-ide/tickets/1001388) is a ticket related to my previous point.

I agree this is not the best design (a good API shouldn't offer sharp knives that you can use to cut yourself), but I don't see a better solution at the moment. Furthermore, consider that other methods in `CompilerControl` also have a `keepLoaded` parameter, so at least we have some consistency.

Bottom line, even though I understand the pain, I don't see how it can be avoided. I think I've been a good boy scout and manually checked that the three PRs I linked work together. This was done using the same uber-build script we use for the Scala validation script. In fact, I just pushed a branch that you can use yourself: https://github.com/dotta/uber-build/tree/transient/test-pr

Just run the following: `./uber-build.sh config/test.conf`

The `config/test.conf` file contains the configuration for building the three PRs together. 

And just to make sure things work as expected, I'm re-running it once more locally and will report back as soon as it's done. (EDIT: I confirm this worked fine).

**Comment by misto** [comment]
@adriaanm I initially had the same reaction - why not keep the them loaded by default - but I now think this is the better way to go, even if it's a bit painful at the moment, in the long run it's the proper solution.

**Comment by adriaanm** [comment]
I don't insist on being a bottleneck here -- was just pitching in until Europe woke up (literally I mean) :-)

@retronym, could you coordinate with @dotta

**Comment by adriaanm** [comment]
Thanks for explaining in detail, @dotta. Sounds reasonable. In general, I'd prefer deprecation cycles and new names when we have to, but I agree we're not there yet.

**Comment by dotta** [comment]
>  I'd prefer deprecation cycles and new names when we have to, but I agree we're not there yet.

But this is a correctness fix. The change is binary compatible, but semantic is affected.

**Comment by adriaanm** [comment]
Also, please preserve your github comments in the commit message

**Comment by adriaanm** [comment]
(The deprecation cycle could let you swap the default for keepLoaded. Forgive my concision, sneakily typing during a meeting.)

**Comment by dotta** [comment]
> Also, please preserve your github comments in the commit message

@adriaanm Done.

**Comment by dotta** [comment]
:wine_glass: thanks!
